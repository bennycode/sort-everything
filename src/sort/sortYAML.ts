import {ParsedNode, parseDocument, YAMLMap, YAMLSeq} from 'yaml';
import type {SortFunction} from './SortFunction';

function sortDeep(node: ParsedNode | null): void {
  if (node instanceof YAMLMap) {
    node.items.sort((itemA, itemB) => (itemA.key < itemB.key ? -1 : itemA.key > itemB.key ? 1 : 0));
    node.items.forEach(item => sortDeep(item.value));
  } else if (node instanceof YAMLSeq) {
    node.items.forEach(item => sortDeep(item));
  }
}

export const sortYAML: SortFunction = function (text: string) {
  try {
    const document = parseDocument(text);
    sortDeep(document.contents);

    return {
      payload: document.toString(),
      type: 'success',
    };
  } catch (error) {
    return {
      error: error as Error,
      type: 'error',
    };
  }
};
