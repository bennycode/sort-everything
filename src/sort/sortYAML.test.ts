import {sortYAML} from './sortYAML';
import assert from 'assert';

describe('sortYAML', () => {
  it('preserves comments when sorting YAML input', () => {
    const input = `
variables:
  C_VARIABLE: 'true' # Comment C
  A_VARIABLE: 'true' # Comment A
  B_VARIABLE: 'true' # Comment B
  `;

    const actual = sortYAML(input);

    assert(actual.type === 'success');
    expect(actual.payload).toMatchSnapshot();
  });
});
