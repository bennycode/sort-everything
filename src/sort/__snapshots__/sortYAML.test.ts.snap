// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortYAML preserves comments when sorting YAML input 1`] = `
"variables:
  A_VARIABLE: 'true' # Comment A
  B_VARIABLE: 'true' # Comment B
  C_VARIABLE: 'true' # Comment C
"
`;
